### ISSUE TITLE
New Check: enforce pattern matching by structs on function parameters

### ISSUE THREAD

**Issue by gabrielperales** [rrrene/credo-proposals#103]
With the upcoming Elixir releases introducing type inference, a rule to enforce pattern matching by struct will be beneficial.

This is the behavior I would expect from this rule:

🛑  Accessing to variable atom property:

```elixir
def get_role(user) do
  user.role # If we are accessing an atom key it should be a Struct.
end
```

🛑   Pattern-matching by an atom key of a map:

```elixir
def get_role(%{role: role}) do # if the map keys are atoms it must be a Struct.
  role
end
```

✅  Pattern-matching by Struct helps the developer know the type and shape of the parameters, it also helps the code editor with autocomplete.

```elixir
def get_role(user = %User{}) do
  user.role
end
```

**Maybe the rule can be split in two**, one enforcing pattern-matching by the property you are using in the function body to avoid access to unexistent fields, and another enforcing using structs when you are pattern-matching by maps with atom fields. 

## First Check 
Prefer pattern-match by map property the body of a function access to a map field:

🛑  Accessing a map field in the body:

```elixir
def get_role(user) do
  user["role"]
end
```

✅   Pattern-matching by an atom key of a map:

```elixir
def get_role(%{"role" => role}) do
  role
end
```

🛑  Accessing a map field in the body:

```elixir
def get_role(user) do
  user.role
end
```

✅   Pattern-matching by an atom key of a map:

```elixir
def get_role(%{role: role}) do
  role
end
```

Maybe it can be called `Credo.Check.Warning.SafeParamAccess`? 🤔 

## Second Check

Prefer pattern-match by struct when pattern-matching by map atom:

🛑   Pattern-matching by an atom key of a map:

```elixir
def get_role(%{role: role}) do # if the map keys are atoms it must be a Struct.
  role
end
```

✅   Pattern-matching by struct

```elixir
def get_role(user = %User{}) do
  user.role
end
```

Maybe it can be called `Credo.Check.Refactor.FunctionMapParams`?  🤔

**Comment by gabrielperales** [comment]
👋 @rrrene I'm implementing these checks, but I would like your opinion. Do you think it makes sense? The second check can be useful for the upcoming Elixir 1.18 to take advantage of the type inference.

**Comment by rrrene** [comment]
Hi,

what would be the benefit of using 

```elixir
def get_role(user = %User{}) do
  user.role
end
```

over

```elixir
def get_role(%User{role: role}) do
  role
end
```

Wouldn't this be equally telling to the compiler?

**Comment by gabrielperales** [comment]
That's not what I said. The benefit is `def get_role(user = %User{}) do` or `def get_role(%User{role: role}) do` over `def get_role(%{role: role}) do`

I want to enforce that if you are pattern-matching by a map with atoms, it probably should be a struct.

**Comment by rrrene** [comment]
> I want to enforce that if you are pattern-matching by a map with atoms, it probably should be a struct.

Okay, I can see how this might be a rule that you could align on in a specific situation.

But why should this be a standard Credo check? People use maps with atom keys all the time or am I misunderstanding something about the proposal?

**Comment by gabrielperales** [comment]
Why would you use a map with atoms instead of a Struct? The only case I can think about is a configuration object/map, but configurations don't have to be passed as parameters to functions because you can access them using `Application.fetch_env!` inside the function, or in a private function (we could disable this rule for private functions).

On the other hand, this rule could help take advantage of [the new type inference coming in Elixir 1.18](https://elixirforum.com/t/elixir-v1-18-0-rc-0-released/68015#p-349274-type-system-improvements-1)

**Comment by peaceful-james** [comment]
I like this rule. I have been trying it out on a side-project and the changes it necessitates introduce a lot of clarity to the code.

Update:
Once place it is not nice is functional components in Phoenix.
I have a lot of them like this:
```
def input(%{errors: nil} = assigns) do
  ~H"""
```

**Comment by gabrielperales** [comment]
> I like this rule. I have been trying it out on a side-project and the changes it necessitates introduce a lot of clarity to the code.
> 
> Update: Once place it is not nice is functional components in Phoenix. I have a lot of them like this:
> 
> ```
> def input(%{errors: nil} = assigns) do
>   ~H"""
> ```

Yes, that's right, for live components this rule should be disabled. Maybe the rule can be configured so that it is not executed on files ending in `live.ex` , or you can disable this rule on those files.

**Comment by rrrene** [comment]
Still, none of this answers the question why this should be a standard check in Credo (and I already said that I think this could be useful for some teams in some situations).

Judging by the linked repo, this was never test-driven in the real world, right?

**Comment by gabrielperales** [comment]
The linked repo is just an example I did for a blog post, but I have a more advanced version used in real-world projects. 

But yeah, maybe it shouldn't be a standard check in Credo, but a plugin or something. 🤔
